zebra-chain/src/amount.rs

```diff
@@ -191,6 +193,94 @@ where
     }
 }
 
+impl<C> Hash for Amount<C>
+where
+    C: Constraint,
+{
+    /// Amounts with the same value are equal, even if they have different constraints
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl<C1, C2> PartialEq<Amount<C2>> for Amount<C1>
+where
+    C1: Constraint,
+    C2: Constraint,
+{
+    fn eq(&self, other: &Amount<C2>) -> bool {
+        self.partial_cmp(other) == Some(Ordering::Equal)
+    }
+}
+
+// We can't implement Eq between different amount constraints,
+// because it leads to an unconstrained type parameter error
+impl<C> Eq for Amount<C>
+where
+    Amount<C>: PartialEq,
+    C: Constraint,
+{
+}
+
+impl<C1, C2> PartialOrd<Amount<C2>> for Amount<C1>
+where
+    Amount<C1>: PartialEq<Amount<C2>>,
+    C1: Constraint,
+    C2: Constraint,
+{
+    fn partial_cmp(&self, other: &Amount<C2>) -> Option<Ordering> {
+        Some(self.0.cmp(&other.0))
+    }
+}
+
+// We can't implement Ord between different amount constraints,
+// because it leads to an unconstrained type parameter error
+impl<C> Ord for Amount<C>
+where
+    Amount<C>: Eq,
+    Amount<C>: PartialOrd,
+    C: Constraint,
+{
+    fn cmp(&self, other: &Amount<C>) -> Ordering {
+        self.partial_cmp(&other).expect("Amount has a total order")
+    }
+}
```

Comment:
I don't think it's possible for `Eq` or `Ord` to compare different types, only PartialEq/PartialOrd can do that since they're the only ones that have a generic parameter and therefore the ability to be implemented multiple times for the same type. I removed all the unnecessary bounds because the Constraintness of C is entirely irrelevant to these trait impls, also rather than implementing each trait in terms of the others I've forwarded everything uniformly to the implementation on the inner `i64` which we can rely upon to be consistent.

```suggestion
impl<C1, C2> PartialEq<Amount<C2>> for Amount<C1> {
    fn eq(&self, other: &Amount<C2>) -> bool {
        self.0.eq(&other.0)
    }
}

impl Eq for Amount<NegativeAllowed> {}
impl Eq for Amount<NonNegative> {}

impl<C1, C2> PartialOrd<Amount<C2>> for Amount<C1> {
    fn partial_cmp(&self, other: &Amount<C2>) -> Option<Ordering> {
        Some(self.0.cmp(&other.0))
    }
}

impl Ord for Amount<NegativeAllowed> {
    fn cmp(&self, other: &Amount<NegativeAllowed>) -> Ordering {
        self.0.cmp(&other.0)
    }
}

impl Ord for Amount<NonNegative> {
    fn cmp(&self, other: &Amount<NonNegative>) -> Ordering {
        self.0.cmp(&other.0)
    }
}
```